packages/modules/providers/locking-redis/src/services/redis-lock.ts

```diff
@@ -0,0 +1,269 @@
+      this.defaultRetryInterval = +options.defaultRetryInterval!
+    }
+
+    if (!isNaN(+options?.maximumRetryInterval!)) {
+      this.maximumRetryInterval = +options.maximumRetryInterval!
+    }
+
+    // Define the custom command for acquiring locks
+    this.redisClient.defineCommand("acquireLock", {
+      numberOfKeys: 1,
+      lua: `
+        local key = KEYS[1]
+        local ownerId = ARGV[1]
+        local ttl = tonumber(ARGV[2])
+        local awaitQueue = ARGV[3] == 'true'
+
+        local setResult = redis.call('SET', key, ownerId, 'EX', ttl, 'NX')
+
+        if setResult then
+          return 1
+        elseif not awaitQueue then
+          -- Key already exists; retrieve the current ownerId
+          local currentOwnerId = redis.call('GET', key)
+          if currentOwnerId == ownerId then
+            redis.call('SET', key, ownerId, 'EX', ttl, 'XX')
+            return 1
+          else
+            return 0
+          end
+        else
+          return 0
+        end
+      `,
+    })
+
+    // Define the custom command for releasing locks
+    this.redisClient.defineCommand("releaseLock", {
+      numberOfKeys: 1,
+      lua: `
+        local key = KEYS[1]
+        local ownerId = ARGV[1]
+
+        if redis.call('GET', key) == ownerId then
+          return redis.call('DEL', key)
+        else
+          return 0
+        end
+      `,
+    })
+  }
+
+  private getKeyName(key: string): string {
+    return `${this.keyNamePrefix}${key}`
+  }
+
+  async execute<T>(
+    keys: string | string[],
+    job: () => Promise<T>,
+    args?: {
+      timeout?: number
+    }
+  ): Promise<T> {
+    const timeout = Math.max(args?.timeout ?? this.waitLockingTimeout, 1)
+    const timeoutSeconds = Number.isNaN(timeout) ? 1 : timeout
+
+    const cancellationToken = { cancelled: false }
+    const promises: Promise<any>[] = []
+    if (timeoutSeconds > 0) {
+      promises.push(this.getTimeout(timeoutSeconds, cancellationToken))
+    }
+
+    promises.push(
+      this.acquire_(
+        keys,
+        {
+          awaitQueue: true,
+        },
+        cancellationToken
+      )
+    )
+
+    await Promise.race(promises)
+
+    try {
+      return await job()
+    } finally {
+      await this.release(keys)
+    }
+  }
+
+  async acquire(
+    keys: string | string[],
+    args?: {
+      ownerId?: string
+      expire?: number
+      awaitQueue?: boolean
+    }
+  ): Promise<void> {
+    return this.acquire_(keys, args)
+  }
+
+  async acquire_(
+    keys: string | string[],
+    args?: {
+      ownerId?: string
+      expire?: number
+      awaitQueue?: boolean
+    },
+    cancellationToken?: { cancelled: boolean }
+  ): Promise<void> {
+    keys = Array.isArray(keys) ? keys : [keys]
+
+    const timeout = Math.max(args?.expire ?? this.waitLockingTimeout, 1)
+    const timeoutSeconds = Number.isNaN(timeout) ? 1 : timeout
+    let retryTimes = 0
+
+    const ownerId = args?.ownerId ?? "*"
```

Comment:
q: Shouldn't this be random when not provided to avoid "imitating owners" or am I missing something?

Reply:
Or alternatively, a required argument

Reply:
Good point, I changed things and the last version is not correct.
It should be "*" or a default value, however in the case the key is present, it should throw instead of extending the ttl.
If we generate a random ownerId, the locking won't be able to be released.
I will fix this in the other providers as well.

Reply:
the `acquire` method can be used for long locks that have `undefined` expiration time.
I realised as well that this is not covered in this provider.